Enhancement: Add +/- syntax, (w)eek (m)onth (y)ear, and time support to query :inputs

bendyorke opened the pull request:
This refactors `db-util/resolve-input` to use a multi method like previously discussed with @logseq-cldwalker. There are several things going on in this refactor:

1. Move all keyword resolvers into distinct multi methods
2. Add support for simple query-like syntax (`+1d`, `-7d`)
3. Add support for relative weeks/months/years (`+1w` `-3m` `+2y`)
4. Marks inputs with older syntax as DEPRECATED (only in the codebase, they just translate to the new inputs and re-resolve)
5. Adds 'support' for other date relatives - currently everything is relative to today, but once this gets merged, I will subsequently add support for the journal date relatives, which should be a single line change (`:journal/+3d`). I also thought this could very easily support `:query/+3d` and `:current/+3d`, but I'm realizing as I'm writing this that the only time those are really applicable are on journal pages, and I'm pretty sure `:journal/+3d` would suffice in both cases. 
    
Let me know if the namespaces are too abstract, I'm open to changing the inputs to something more readable like `+3d-journal` or `:+3d-journal-day` or `+3d-from-journal-day`, but none of those read right to me. I'm not sure if namespaces are too confusing, but since the day that the query is relative to is kind of like a context, and the rest should function the same, it seemed fitting.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/logseq/logseq?pullRequest=8387) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/logseq/logseq?pullRequest=8387) before we can accept your contribution.<br/><hr/>**Ben Yorke** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/logseq/logseq?pullRequest=8387) it.</sub>

bendyorke commented on deps/graph-parser/src/logseq/graph_parser/util/db.cljs:
```diff
@@ -22,56 +22,93 @@ it will return 1622433600000, which is equivalent to Mon May 31 2021 00 :00:00."
+        ms? (string/ends-with? (name input) "-ms")]
+    (keyword :today (str plus-minus count "d" (if ms? "-ms" "")))))
+
+(comment
```
I'm not sure whether to remove this block, keep it, or move it to tests.

We are still testing the old date formats, so it seems redundant to write tests for this function

bendyorke commented on deps/graph-parser/src/logseq/graph_parser/util/db.cljs:
```diff
@@ -22,56 +22,93 @@ it will return 1622433600000, which is equivalent to Mon May 31 2021 00 :00:00."
+  (date-time-util/time-ms))
+
+(defmethod resolve-keyword-input :relative-date [db input opts]
+  (let [relative-to (case (or (namespace input) "today")
```
This case looks strange now, but shortly will add a `"journal"` case, and in the future could add things like `"prop-.*"` case to have queries be relative to a block property 

so something like the following would resolve relative to 20230101, and you could change the prop instead of mucking with the query:

```
date:: 20230101
#+BEGIN_QUERY 
{:name "test"
 :inputs [:prop-date/-3d :prop-date/+3d]
 :query [:find (pull ?b [*])
         :in $ ?before ?after 
         (between ?b ?before ?after)]}
#+END_QUERY
```

bendyorke replied:
Could also actually support `query` and `current` as I originally anticipated - if people are embedded blocks across journal pages ¯\_(ツ)_/¯

logseq-cldwalker replied:
case makes sense for now. I'm not sure if properties as part of the input syntax would be easy to support as properties can have namespaces and pretty much anything that is valid edn including whitespace

bendyorke replied:
I think full support for properties then may be a challenge - but to me it would be satisfactory if it only did a match on the name, and would fail if there was anything besides a date type object in the value

logseq-cldwalker commented on deps/graph-parser/src/logseq/graph_parser/util/db.cljs:
```diff
@@ -22,56 +22,93 @@ it will return 1622433600000, which is equivalent to Mon May 31 2021 00 :00:00."
+(defmethod resolve-keyword-input :right-now-ms [_ _ _]
+  (date-time-util/time-ms))
+
+(defmethod resolve-keyword-input :relative-date [db input opts]
```
The first couple times I read this I was confused where the ms handling went as I only thought of journal date ints being here. Could we give this a more explicit name like `:relative-date-and-time` or have separate defmethods for date and time?

bendyorke replied:
yeah that's a much better name indeed 🔥

bendyorke replied:
I think, at least at the moment, the date and time functions combine nicely, as 90% of their logic is shared

bendyorke replied:
Moved these out actually into separate `:relative-date` and `:relative-date-time` resolvers since I added some better (more explicit) support for time manipulation (`start` `end` and `HHMM` in addition to the dynamic `ms`)

logseq-cldwalker commented on deps/graph-parser/src/logseq/graph_parser/util/db.cljs:
```diff
@@ -22,56 +22,93 @@ it will return 1622433600000, which is equivalent to Mon May 31 2021 00 :00:00."
+    (re-find #"^\d+d(-before|-after|-before-ms|-after-ms)?$" (name input)) :DEPRECATED-relative-date
+    (= :start-of-today-ms input) :DEPRECATED-start-of-today-ms 
+    (= :end-of-today-ms input) :DEPRECATED-end-of-today-ms))
+(defmulti resolve-keyword-input (fn [_db input _opts] (keyword-input-dispatch input))) 
```
❤️

logseq-cldwalker commented on deps/graph-parser/src/logseq/graph_parser/util/db.cljs:
```diff
@@ -22,56 +22,93 @@ it will return 1622433600000, which is equivalent to Mon May 31 2021 00 :00:00."
+  (resolve-keyword-input db (old->new-relative-date-format input) opts))
+
+(defmethod resolve-keyword-input :DEPRECATED-start-of-today-ms [db _ opts]
+  (resolve-keyword-input db :today/-0d-ms opts))
```
I don't know why -0d-ms and +0d-ms aren't intuitive to me. Maybe it's worth writing these in hour form or not deprecating these two?

bendyorke replied:
Yeah I didn't know what to do about these keywords. Since the logic is the same as `:today/-0d-ms`, I thought it best to keep the logic in one place instead of just duplicate it. 

Maybe deprecated isn't the right word, ALIASED? The only reason to indicate anything is 'performance' as this input will re-resolve (I don't think resolving a single input though is going to be a real performance hit). Maybe I can remove any label?

logseq-cldwalker replied:
If we're not meaning to deprecate (to devs), then removing or ALIASED makes more sense

bendyorke replied:
I took these out of deprecated, now they're all going to a `:today-time` resolver :)

logseq-cldwalker commented on deps/graph-parser/src/logseq/graph_parser/util/db.cljs:
```diff
@@ -22,56 +22,93 @@ it will return 1622433600000, which is equivalent to Mon May 31 2021 00 :00:00."
+(defmethod resolve-keyword-input :relative-date [db input opts]
+  (let [relative-to (case (or (namespace input) "today")
+                      "today" (t/today))
+        [_ offset-direction offset offset-unit offset-ms?] (re-find #"^([+-])(\d+)([dwmy])(-ms)?$" (name input))
```
❤️ how this is clean and concise all the logic is here

logseq-cldwalker commented on src/test/frontend/db/query_react_test.cljs:
```diff
@@ -129,3 +153,92 @@ created-at:: %s"
                                         [(<= ?timestamp ?end)]]})))
         ":Xd-before-ms and :Xd-after-ms resolve to correct datetime range")))
+
+(deftest resolve-input-for-relative-date-queries 
```
🔥

logseq-cldwalker commented on src/test/frontend/db/query_react_test.cljs:
```diff
@@ -106,6 +106,30 @@ created-at:: %s"
+                                        [(<= ?timestamp ?end)]]})))
+        ":+0d-ms and :-0d-ms resolve to correct datetime range")
+
+    (is (= ["yesterday" "today"]
```
Optional: It could be nice to have tests for w, m, y for ms

logseq-cldwalker replied:
Thanks for adding these!

logseq-cldwalker commented on deps/graph-parser/src/logseq/graph_parser/util/db.cljs:
```diff
@@ -22,56 +22,93 @@ it will return 1622433600000, which is equivalent to Mon May 31 2021 00 :00:00."
+  (let [relative-to (case (or (namespace input) "today")
+                      "today" (t/today))
+        [_ offset-direction offset offset-unit offset-ms?] (re-find #"^([+-])(\d+)([dwmy])(-ms)?$" (name input))
+        offset-fn (case offset-direction "+" t/plus "-" t/minus)
```
I'm digging the explicitness of the syntax we agreed to. This sets us up nicely in the future to support chaining these down the road :)

bendyorke replied:
What do you mean by chaining these?

logseq-cldwalker replied:
Don't want to scope creep this PR hehe, but down the road we could get asks to add or subtract multiple date ranges so `+1w+1d` or `+1y-1w` would naturally fit

logseq-cldwalker reviewed (changes requested):
@bendyorke Great work on these input enhancements with date-time math! Back to you for some feedback and to get CI to green

> Let me know if the namespaces are too abstract, I'm open to changing the inputs to something more readable like +3d-journal or :+3d-journal-day or +3d-from-journal-day, but none of those read right to me.

I find the emphasis on date math first more readable as suggested by https://github.com/logseq/logseq/discussions/7866. But maybe having the terser +/- changes that. How about we ask non-engineers about it? If we do decide to put the date-time type first, I'd advocate for `-` as the separator over `/` as the latter seems less readable

logseq-cldwalker commented:
@bendyorke Changed title since PR titles get used in changelogs so we try to make them as readable for users as possible. Feel free to reword

bendyorke commented on deps/graph-parser/src/logseq/graph_parser/util/db.cljs:
```diff
@@ -22,56 +22,93 @@ it will return 1622433600000, which is equivalent to Mon May 31 2021 00 :00:00."
+  (some-> (current-page-fn) string/lower-case))
+
+(defmethod resolve-keyword-input :current-block [db _ {:keys [current-block-uuid]}]
+  ;; TODO: handle current-block-uuid not being provided
```
Ah I thought this would fail

logseq-cldwalker replied:
I thought `d/entity` would return nil but feel free to double check

bendyorke commented on deps/graph-parser/src/logseq/graph_parser/util/db.cljs:
```diff
@@ -22,56 +22,93 @@ it will return 1622433600000, which is equivalent to Mon May 31 2021 00 :00:00."
+  (resolve-keyword-input db :today/+0d-ms opts))
+
+(defmethod resolve-keyword-input :current-page [_ _ {:keys [current-page-fn]}]
+  ;; TODO: handle current-page-fn not being provided
```
👍 if error handling isn't outside the scope of this PR, then I will add it here! I didn't want to do too many things, already doing a lot haha

logseq-cldwalker replied:
I was thinking of just the `:or` used in the parent fn (which is inconsistent with :current-block-uuid now that I think about it) but if you're doing error handling too that's great. I'd also be fine with just throwing an explicit error to console as these are handling unexpected errors. If you want to do something more user facing, something like `ui/block-error` could be handy

logseq-cldwalker replied:
Thanks for addressing these

bendyorke commented on deps/graph-parser/src/logseq/graph_parser/util/db.cljs:
```diff
@@ -22,56 +22,93 @@ it will return 1622433600000, which is equivalent to Mon May 31 2021 00 :00:00."
+    (try 
+      (resolve-keyword-input db input {:current-block-uuid current-block-uuid
+                                       :current-page-fn current-page-fn})
+      (catch js/Error e
```
consistency was the only motivation here, but I'll add proper error handling since that's not out of scope :)

logseq-cldwalker replied:
Thanks for addressing

bendyorke commented:
New PR title is better than I could have come up with 😂 good to know they're used in changelogs, I'll try to be more explicit in the future. Please feel free to rename anytime you'd like a change, I appreciate it!

bendyorke commented:
@logseq-cldwalker I think I've addressed all of your concerns. LMK if there's anything else, once this get's merged in I'll update the docs graph with info about these :D

logseq-cldwalker commented:
> @logseq-cldwalker I think I've addressed all of your concerns. LMK if there's anything else, once this get's merged in I'll update the docs graph with info about these :D

Sweet. Taking a look. Can you get CI to green in the meantime? There's a new version of `@logseq/nbb-logseq` for the related lint task

logseq-cldwalker commented on src/test/frontend/db/query_react_test.cljs:
```diff
@@ -39,6 +39,16 @@ adds rules that users often use"
+                                      :where [?b :block/page ?bp]
+                                      [?bp :block/name ?current-page]]}
+                            {:current-page-fn nil})))
+      ":current-page input fails gracefully when not present")
```
I found this confusing as it's not failing but rather not resolving the input. I'm happy to tweak this wording later

logseq-cldwalker commented on deps/graph-parser/src/logseq/graph_parser/util/db.cljs:
```diff
@@ -36,37 +36,68 @@ it will return 1622433600000, which is equivalent to Mon May 31 2021 00 :00:00."
-  ;; :Xd, :Xd-before, :Xd-before-ms, :Xd-after, :Xd-after-ms
-  (resolve-keyword-input db (old->new-relative-date-format input) opts))
+    (re-find #"^[+-]\d+[dwmy]?$" (name input)) :relative-date
+    (re-find #"^[+-]\d+[dwmy]-(ms|start|end|\d{2}|\d{4}|\d{6}|\d{9})?$" (name input)) :relative-date-time
```
Thinking more on this, this doesn't need to block this PR. We can adjust the syntax before release as needed

logseq-cldwalker reviewed (approved):
@bendyorke Thanks for all the great query :inputs features you've added! 👍 🙌 :shipit:   Looks like there's just a nbb-logseq bump to get the last test passing. I can do that in a bit if you don't get to it
